### ISSUE TITLE
Rule proposal: prefer hash names for private members

### ISSUE THREAD

**Issue by kimamula** [typescript-eslint/typescript-eslint#1391]
TypeScript 3.8 will add support for private named instance fields. What about creating a rule for disallowing the use of the `private` accessibility modifier in favor of the standard hash names for private members?

Examples of **incorrect** code for this rule:

```ts
class MyClass {
  private foo = 1;
  constructor(private bar: string) {}
}
```

Examples of **correct** code for this rule:

```ts
class MyClass {
  #foo = 1;
  #bar: string;
  constructor(bar: string) {
    this.#bar = bar;
  }
}
```

### Options

```ts
// Any target whose hash names are supported
// by the TypeScript version in use can be specified
type Target =
//  | 'accessors'
//  | 'constructors'
//  | 'staticMethods'
//  | 'instanceMethods'
//  | 'staticFields'
  | 'instanceFields';

type Options = Target[];

// Defaults to all the targets
const defaultOptions: Options = ['instanceFields']; // for TypeScript 3.8
```

**Comment by glen-84** [comment]
Perhaps it should allow the opposite as well – enforcing soft (compile-time-only) privacy?

**Comment by bradzacher** [comment]
Probably best to be added as an option on the `explicit-member-accessibility` rule.

**Comment by kimamula** [comment]
Then the options for `explicit-member-accessibility` should be changed to something like the following?

```ts
type AccessibilityLevel =
  | 'explicit' // require an accessor (including public)
  | 'no-public' // don't require public
  | 'off'; // don't check

type AvailableModifier = 'public' | 'protected' | 'private' | '#';

type AccessibilitySetting = AccessibilityLevel | AvailableModifier[];

type Options = {
  accessibility?: AccessibilitySetting;
  ignoredMethodNames?: string[];
  overrides?: {
    accessors?: AccessibilitySetting;
    constructors?: AccessibilitySetting;
    methods?: AccessibilitySetting | {
      static?: AccessibilitySetting;
      instance?: AccessibilitySetting;
    };
    properties?: AccessibilitySetting | {
      static?: AccessibilitySetting;
      instance?: AccessibilitySetting;
    };
    parameterProperties?: AccessibilitySetting;
  };
};

const defaultOptions: Options = {
  accessibility: 'explicit',
};
```

I'm afraid that the name `explicit-member-accessibility` is a bit confusing for this use case though.
Perhaps it should be renamed to something like `member-accessibility` and the name `explicit-member-accessibility` should remain as an alias for it?

**Comment by zheeeng** [comment]
Expect more discussions and progress on this issue.

**Comment by bradzacher** [comment]
With any issue in opened in this project - it either has a visible progress in the form of an attached PR, or it has no progress.

We are a community run project. The **volunteer** maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues *will not progress* unless a member of the community steps up and **champions** it.

If this issue is important to you - consider being that champion.

If not - please just subscribe to the issue and wait patiently.
Commenting asking for status updates does not bump issue priority in any way and just serves to spam ***everyone*** subscribed to the issue.

**Comment by npdev453** [comment]
Current workaround:

```ts
'@typescript-eslint/naming-convention': [
  'error',
  {
    selector: ['property'],
    modifiers: ['private'],
    prefix: ['#'],
    format: null,
  }
],
```

**Comment by bradzacher** [comment]
As per our [FAQ article](https://typescript-eslint.io/docs/linting/troubleshooting#how-can-i-ban-specific-language-feature) - this can easily be achieved using `no-restricted-syntax`:
```
no-restricted-syntax: [
    'error',
    {
        selector: ':matches(PropertyDefinition, MethodDefinition)[accessibility="private"]',
        message: 'Use #private instead',
    },
],
```
[playground](https://typescript-eslint.io/play/#ts=4.6.4&sourceType=module&showAST=es&code=PQKgBApgzgNglgOwC5gQewLQCdpK3AYyQgBMMoBPZAQwA8AuMAbQCgx2wByCLLNLTgBo2HAN4iOHKBBgQi-Rp3oBbakgIALaAAoACnwAOPJBQAiEAGaI4SOGgSCwAWQhINaEuasIbdhAEomagICaCg4ACM4eBMAXgAiA3wANzUIeIBdIQlJMGUw6gBzCEUAVWkwAGIkuFTiMEQoYmoSbMkAX2EMsBBgFhYCGGooKDAAMTQ0MHFJJLQDAEZGagQKAG4cgwBXCPgCMDmDACZl1Y3ZvmIiUgPDAGZT9c2UtNv5gBZH845qwwBWL79ST5NweBbafzTdqbHZ7PKudwkI4QqHPNBXYgkeGgkh3FGiaEXWqvEGI974wk-UkeP4UoEcYooYKhEb8cGQ0RgHBILZYBBgBZgNaU9jbXaEMCMsDMsL8ZEcrmuXn8wXCtEYm5SmWsrB4hXc5UCoUi27E+pakKyrDk-VKvlGtWSKWVbVQfi0208+2q6HQoA&rules=N4XyA&tsConfig=N4XyA)

similarly the inverse can easily be enforced as well:
```
no-restricted-syntax: [
    'error',
    {
        selector: ':matches(PropertyDefinition, MethodDefinition) > PrivateIdentifier.key',
        message: 'Use `private` instead',
    },
]
```
[playground](https://typescript-eslint.io/play/#ts=4.6.4&sourceType=module&showAST=es&code=PQKgBApgzgNglgOwC5gQewLQCdpK3AYyQgBMMoBPZAQwA8AuMAbQCgx2wByCLLNLTgBo2HAN4iOHKBBgQi-Rp3oBbakgIALaAAoACnwAOPJBQAiEAGaI4SOGgSCwAWQhINaEuasIbdhAEowAD4wfTgANzUIAEkSCGQ4Kx4AOgBrCAohCUkwZWgoagBzCEUAVWkwAAMDfEjiSrBEKGJqEizJAF9hAF0wEGAWFgIYaigoMAAxNDQwcUkatAMARkZqBAoAbmyDAFcAI3gCMAWDACZV9a35vmIiUmPDAGYLze3aqIfFgBYXq44AYhOAFZfoNJHk3B4ltpAqIOtt9odcq53CRTjDZvDrmhbsQSMjISRHhi4W8Ih8IaiviSsQDKR4gTSwRxiihqAQCPl+NDYWAcEgdlgEGAlmANrT2LsDoQwKywOzOWN+OjefzBcLReK3ji5HjZa55RyuVhiarXOqRWKJQ9ycR9WyjUqsNSzQKhZatZI5f8FcbGa6LZr4fCgA&rules=N4XyA&tsConfig=N4XyA)

I don't believe that a lint rule makes sense here. We want to avoid adding rules which just ban a specific language features unless we can add some real additional value. Given that there's no fixer we can provide (because it's unsafe), and there's no options to configure or tweak - I don't see any additional value we can provide.

For this reason, we shall reject this rule proposal.

**Comment by thetutlage** [comment]
> I don't believe that a lint rule makes sense here. We want to avoid adding rules which just ban a specific language features unless we can add some real additional value

I don't think we are banning language features here. Infact, we are asking to promote the language feature ie private properties

**Comment by bradzacher** [comment]
> I don't think we are banning language features here. Infact, we are asking to promote the language feature ie private properties

What's the difference to a linter?
In reality - there's no difference. A lint rule will report on any class member that uses `private`. There's no way to subtly suggest - it will just be a lint error on every single private member.

In practice this is exactly the same as "banning" `private` altogether.

**Comment by thetutlage** [comment]
`private` is a modifier and `#` is a private property. Both are different things all together.

All we are asking is to enforce the user of `#` in favor of `private`. 

> In practice this is exactly the same as "banning" private altogether.

Don't we do that already in other places like? 

- Ban double quotes in favor of single quotes
- Ban usage of `var` in favor of `const` and `let`

**Comment by bradzacher** [comment]
Sorry, I think you've misunderstood. My [comment above hives you a solution to do exactly what you want](https://github.com/typescript-eslint/typescript-eslint/issues/1391#issuecomment-1124154589)

Why would we build a separate rule when there is already existing tooling that more than meets the requirements?
